source/http2_stream_manager.c

```diff
@@ -0,0 +1,801 @@
+#undef DEBUG
+
+#define STREAM_MANAGER_LOGF(level, stream_manager, text, ...)                                                          \
+    AWS_LOGF_##level(AWS_LS_HTTP2_STREAM_MANAGER, "id=%p: " text, (void *)(stream_manager), __VA_ARGS__)
+#define STREAM_MANAGER_LOG(level, stream_manager, text) STREAM_MANAGER_LOGF(level, stream_manager, "%s", text)
+
+static void s_stream_manager_start_destroy(struct aws_http2_stream_manager *stream_manager);
+static void s_aws_http2_stream_manager_execute_transaction(struct aws_http2_stream_management_transaction *work);
+
+static struct aws_h2_sm_pending_stream_acquisition *s_new_pending_stream_acquisition(
+    struct aws_allocator *allocator,
+    const struct aws_http_make_request_options *options,
+    aws_http2_stream_manager_on_stream_acquired_fn *callback,
+    void *user_data) {
+    struct aws_h2_sm_pending_stream_acquisition *pending_acquisition =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_h2_sm_pending_stream_acquisition));
+
+    /* Copy the options and keep the underlying message alive */
+    pending_acquisition->options = *options;
+    pending_acquisition->request = options->request;
+    aws_http_message_acquire(pending_acquisition->request);
+    pending_acquisition->callback = callback;
+    pending_acquisition->user_data = user_data;
+    pending_acquisition->allocator = allocator;
+    return pending_acquisition;
+}
+
+static void s_pending_stream_acquisition_destroy(struct aws_h2_sm_pending_stream_acquisition *pending_acquisition) {
+    if (pending_acquisition == NULL) {
+        return;
+    }
+    if (pending_acquisition->request) {
+        aws_http_message_release(pending_acquisition->request);
+    }
+    aws_mem_release(pending_acquisition->allocator, pending_acquisition);
+    return;
+}
+
+/**
+ * Main actions:
+ * - Acquire stream from user
+ *      - Check a connection to make stream
+ *          - Schedule task from connection eventloop to make request.
+ *      - Or make a new connection
+ *          - Acquiring a new connection and wait until CM to finish.
+ * - Stream completed from HTTP
+ *      - Update the connection of how many streams opening
+ *      - Connection may be back to available.
+ * - Connection acquired from CM
+ *      - Finishes the pending stream acquiring on this connection as much as possible
+ */
+
+static void s_lock_synced_data(struct aws_http2_stream_manager *stream_manager) {
+    int err = aws_mutex_lock(&stream_manager->synced_data.lock);
+    AWS_ASSERT(!err && "lock failed");
+    (void)err;
+}
+
+static void s_unlock_synced_data(struct aws_http2_stream_manager *stream_manager) {
+    int err = aws_mutex_unlock(&stream_manager->synced_data.lock);
+    AWS_ASSERT(!err && "unlock failed");
+    (void)err;
+}
+
+/* *_synced should only be called with LOCK HELD or from another synced function */
+static bool s_aws_http2_stream_manager_should_destroy_synced(struct aws_http2_stream_manager *stream_manager) {
```

Comment:
there's a lot of stuff here named "should destroy" and I found it confusing as a newcomer
maybe a naming convention like:
shutting_down_1_waiting_on_streams
shutting_down_2_waiting_on_conn_mgr
???

I don't know, just throwing out ideas

Reply:
The should destroy only true when nothing opening on stream manager besides the connection manager.